src/main/java/com/connexta/store/controllers/StoreController.java

```diff
@@ -129,6 +109,111 @@
+    InputStream inputStream;
+    InputStream metacardInputStream;
+    try {
+      inputStream = file.getInputStream();
+      metacardInputStream = metacard.getInputStream();
+    } catch (IOException e) {
+      throw new ValidationException("Could not open attachment"); // TODO Replace this exception
+    }
+    storeService.ingest(
+        file.getSize(),
+        file.getContentType(),
+        inputStream,
+        fileName,
+        metacard.getSize(),
+        metacardInputStream);
+
+    return ResponseEntity.accepted().build();
+  }
+
+  @ApiOperation(
+      value = "Get a metacard for a dataset.",
+      nickname = "retrieveMetacard",
+      response = Resource.class,
+      tags = {"store"})
+  @ApiResponses(
+      value = {
+        @ApiResponse(code = 200, message = "Get Metacard", response = Resource.class),
+        @ApiResponse(
+            code = 401,
+            message = "The client could not be authenticated. ",
+            response = ErrorMessage.class),
+        @ApiResponse(
+            code = 400,
+            message =
+                "The client message could not be understood by the server due to invalid format or syntax. ",
+            response = ErrorMessage.class),
+        @ApiResponse(
+            code = 403,
+            message = "The client does not have permission. ",
+            response = ErrorMessage.class),
+        @ApiResponse(
+            code = 501,
+            message = "The requested API version is not supported and therefore not implemented. ",
+            response = ErrorMessage.class)
+      })
+  @RequestMapping(
+      value = RETRIEVE_METACARD_URL_TEMPLATE,
+      produces = {MediaType.APPLICATION_OCTET_STREAM_VALUE, MediaType.APPLICATION_XML_VALUE},
+      method = RequestMethod.GET)
+  public ResponseEntity<Resource> retrieveMetacard(
+      @Pattern(regexp = "^[0-9a-zA-Z]+$")
+          @Size(min = 32, max = 32)
+          @ApiParam(value = "The ID of the dataset. ", required = true)
+          @PathVariable("datasetId")
+          final String datasetId) {
+    InputStream inputStream = null;
+    try {
+      // TODO return 404 if key doesn't exist
+      inputStream = storeService.retrieveMetacard(datasetId);
+      log.info("Successfully retrieved metacard from datasetId={}", datasetId);
+      return ResponseEntity.ok()
+          .contentType(METACARD_MEDIA_TYPE)
+          .body(new InputStreamResource(inputStream));
+    } catch (RuntimeException e) {
+      if (inputStream != null) {
+        try {
+          inputStream.close();
+        } catch (IOException ioe) {
+          log.warn(
+              "Unable to close InputStream when retrieving metacard from datasetId={}",
+              datasetId,
+              ioe);
+        }
+      }
+
+      log.warn("Unable to retrieve metacard datasetId={}", datasetId, e);
+      throw new StoreException(String.format("Unable to retrieve metacard: %s", e.getMessage()), e);
```

Comment:
❓ Should we really be catching Throwable here?

Reply:
Life is wild

Reply:
Yes. We have to catch throwable to make sure that the input stream gets closed. I don't know of another solution.